sw/device/lib/crypto/include/datatypes.h

```diff
@@ -23,23 +24,41 @@ extern "C" {
 /**
  * Enum to handle return values of the crypto API.
  *
- * Values are hardened.
+ * Values are built to be bit-compatible with OpenTitan's internal `status_t`
+ * datatypes. The highest (sign) bit indicates if the value is an error (1) or
+ * not (0). For non-error statuses, the rest can be anything; in cryptolib
+ * status codes it is always `kHardenedBoolTrue`. For errors:
+ *   - The next 15 bits are a module identifier, which is always 0 in the
+ *     cryptolib status codes
+ *   - The next 11 bits are a line number or other information; in the
+ *     cryptolib status codes, it is a hardened value created to have high
+ *     Hamming distance with the other valid status codes
+ *   - The final 5 bits are an Abseil-compatible error code
+ *
+ * The hardened values for error codes were generated with:
+ * $ ./util/design/sparse-fsm-encode.py -d 5 -m 5 -n 11 \
+ *      -s 4232058530 --language=sv --avoid-zero
+ *
+ * Use the same seed value and a larger `-m` argument to generate new values
+ * without changing all error codes. Remove the seed (-s argument) to generate
+ * completely new 11-bit values.
  */
-typedef enum crypto_status {
+typedef status_t crypto_status_t;
+typedef enum crypto_status_value {
   // Status is OK; no errors.
-  kCryptoStatusOK = 0x739,
+  kCryptoStatusOK = (int32_t)0x739,
   // Invalid input arguments; wrong length or invalid type.
-  kCryptoStatusBadArgs = 0xb07,
+  kCryptoStatusBadArgs = (int32_t)0x8000fea3,
   // Error after which it is OK to retry (e.g. timeout).
-  kCryptoStatusInternalError = 0x5c3,
+  kCryptoStatusInternalError = (int32_t)0x8000534a,
   // Error after which it is not OK to retry (e.g. integrity check).
-  kCryptoStatusFatalError = 0xf5c,
+  kCryptoStatusFatalError = (int32_t)0x80006d89,
   // An asynchronous operation is still in progress.
-  kCryptoStatusAsyncIncomplete = 0xae1,
+  kCryptoStatusAsyncIncomplete = (int32_t)0x8000ea4e,
   // TODO: remove all instances of this error before release; it is to track
   // implementations that are not yet complete.
-  kCryptoStatusNotImplemented = 0xff,
-} crypto_status_t;
+  kCryptoStatusNotImplemented = (int32_t)0x80008d2c,
```

Comment:
Consider writing these as:
```
kCryptoStatusInternalError = (int32_t)0x80005340 | kAborted,
...etc...
```

It appears that they all have the correct bit pattern in the low bits, but it isn't completely obvious without pulling up `absl_status.h` and cross checking :)

Reply:
Done, and also added tests to both debug/non-debug unit tests to check that the extracted errors are as expected.